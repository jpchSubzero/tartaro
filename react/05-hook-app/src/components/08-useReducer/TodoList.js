import React from 'react';
import { TodoListItem } from './TodoListItem';

export const TodoList = ({todos, handleToggle, handleDelete}) => {
    return (
        <ul>
            {
                todos.length > 0 ?
                    todos.map((todo, idx) => 
                        // Agregar el key para quitar el warning
                        <TodoListItem key={todo.id} todo={todo} idx={idx} handleToggle={handleToggle} handleDelete={handleDelete}/>
                    )
                :
                    <h3>No hay tareas</h3>
            }
        </ul>    
    )
}
