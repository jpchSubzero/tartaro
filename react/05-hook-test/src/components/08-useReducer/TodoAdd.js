import React from 'react';
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({handleAddTodo}) => {
    const [{ description }, handleInputChange, handleReset] = useForm({
        description: ''
    });

    const handleSubmit = (e) => {
        e.preventDefault();
        if (description.trim().length < 3) {
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false            
        };

        handleAddTodo(newTodo);
        handleReset();
    };    

    return (
        <>
            <div className="col-5">
                <h4>Agregar tareas</h4>
                <hr />

                <form className="form-group" onSubmit={ handleSubmit }>
                    <input 
                        type="text" 
                        name="description" 
                        placeholder="Ingrese la tarea" 
                        autoComplete="off" 
                        className="form-control" 
                        onChange={ handleInputChange } 
                        value={ description } 
                        minLength="3"
                    />
                    <button type="submit" className="btn btn-outline-primary mt-2 w-100">Agregar</button>
                </form>
            </div>
        </>
    )
}
