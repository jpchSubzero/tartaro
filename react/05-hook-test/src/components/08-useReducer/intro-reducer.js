const initialTodos = [{
    id: 1,
    todo: 'Comprar pan',
    done: false
}];

const todoReducer = (state = initialTodos, action) => {
    switch (action?.type) {
        case 'add':
            return [...state, action.payload];
        default:
            break;
    }

    return state;
}

let todos = todoReducer();

const milkTodo = {
    id: 2,
    todo: 'Comprar leche',
    done: false
};

const addAction = {
    type: 'add',
    payload: milkTodo
};

todos = todoReducer(todos, addAction);

