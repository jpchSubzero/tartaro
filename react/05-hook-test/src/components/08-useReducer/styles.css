body {
    padding: 70px;
}

/* Estilo para que aparezca la mano y saber si se puede hacer clic. El margen es para que no se separe mucho del borde */
p {
    cursor: pointer;
    margin: 0px;
}

/* Estilo para que el item de la lista y el botón estén seguidos, no en líneas separadas, y con un espacio entre ellos */
.list-group-item {
    align-items: center;
    display: flex;
    justify-content: space-between;
}

/* Estilo para que cuando la tarea esté hecha la tache */
.complete {
    text-decoration: line-through;
}