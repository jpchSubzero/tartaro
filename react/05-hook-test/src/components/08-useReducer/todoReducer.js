export const todoReducer = (state = [], action) => {
    // ? por si no se envía action
    switch (action?.type) {
        // Acción de agregar
        case 'add':
            return [...state, action.payload];

        // Acción de eliminar una tarea
        // Esta acción no va a eliminar realmente el elemento sino a devolver una lista sin ese elemento ya que por la inmutabilidad del state no se puede/debe modificarlo
        // El payload puede ser un objeto todo y utilizar el id o directamente enviar solo el id, como no hay un tipado fuerte no importa si se utilizan otros tipos
        case 'delete':
            return state.filter(todo => todo.id !== action.payload);

        // Manejador de cambio de estado de tarea
        case 'toggle':
            // // Forma larga
            // return state.map(todo => {
            //     if (todo.id === action.payload) {
            //         return {
            //             ...todo,
            //             done: !todo.done,
            //         };
            //     } 
            //     return todo;
            // });

            // Forma corta
            return state.map(todo => todo.id === action.payload ? { ...todo, done: !todo.done } : todo);


        // Acción por defecto que sería devolver el state
        default:
            return state;
    }
}
