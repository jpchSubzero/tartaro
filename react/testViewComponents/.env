REACT_APP_PRODUCT_BASE_URL=https://des-api-eva.novaecuador.com/product/api/product/
REACT_APP_PRODUCT_CHANNEL_CODE=BANCO01
REACT_APP_PRODUCT_INSURANCE_TYPE_CODE=RAM03
REACT_APP_PRODUCT_INSURANCE_CARRIER_CODE=ASE02
REACT_APP_PRODUCT_CONTRACT_TYPE_CODE=MASS_INSURANCE
REACT_APP_PRODUCT_PRODUCT_CODE=PROD005
REACT_APP_PRODUCT_PLAN_CODE=PLAN012
REACT_APP_PRODUCT_WAY_CODE=BANCAWEB
REACT_APP_PRODUCT_PAYMENT_PERIODICITY_CODE=MONTHLY
REACT_APP_PRODUCT_PAYMENT_METHOD_CODE=CREDIT_CARD
REACT_APP_SUBSCRIPTION_BASE_URL=https://des-api-eva.novaecuador.com/subscription/api/subscription/
REACT_APP_SALE_BASE_URL=https://des-api-eva.novaecuador.com/sale/api/sale/
REACT_APP_INTEGRATION_BASE_URL=https://des-api-eva.novaecuador.com/integration/api/integration/
REACT_APP_MESSENGER_SERVICE_BASE_URL=https://des-api-eva.novaecuador.com/messengerservice/api/messengerservice/
REACT_APP_FLOW_MANAGER_BASE_URL=https://des-api-eva.novaecuador.com/flowmanager/api/
