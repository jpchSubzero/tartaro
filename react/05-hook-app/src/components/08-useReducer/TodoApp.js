import React, { useEffect, useReducer } from 'react';
import {todoReducer} from './todoReducer';

// Importar el custom hook useForm para manejar el formulario
// Mover a TodoAdd
// import { useForm } from '../../hooks/useForm';

import './styles.css';

// Solución Tarea 3
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';

export const TodoApp = () => {

    // Luego de definir el useReducer, definir el initialState. El reducer se lo crea en un archivo aparte aunque no es necesario hacerlo así.
    // Luego de la validación del campo pasamos esta parte de inicialización a la función init para hacer lo mismo pero de otra forma
    // const initialState = [{
    //     id: new Date().getTime(),
    //     desc: 'Comprar pan',
    //     done: false
    // }];

    // Definición de useReducer
    // const [state, dispatch] = useReducer(reducer, initialState, init);

    // useState sin dispatch ni init. Cambiamos reducer por el que definimos aparte, en este caso todoReducer
    // const [state] = useReducer(todoReducer, initialState);

    // Cambiamos el nombre de state por lo que representa para que sea más fácil de entender. Como es posicional el nombre no es importante
    // const [todos] = useReducer(todoReducer, initialState);

    // Agregar dispatch cuando se vaya a hacer la acción de agregar
    // const [todos, dispatch] = useReducer(todoReducer, initialState);
    
    // Agregamos la función de init para mejorar el rendimiento. Cambiamos el initialState por [] ya que estamos trabajando con un arreglo de tareas
    const init = () => {
        // return [{
        //     id: new Date().getTime(),
        //     desc: 'Comprar pan',
        //     done: false
        // }];        

        // Una vez realizada la persistencia en el localStorage ya no necesitamos tener quemado un valor inicial sino obtener lo que haya almacenado
        // Esto funciona pero tiene el problema de la primera vez, es decir, cuando no hay nada va a devolver null por lo que se debe controlar haciendo que devuelva []
        // return JSON.parse(localStorage.getItem('todos'));

        return JSON.parse(localStorage.getItem('todos')) || [];
    };

    const [todos, dispatch] = useReducer(todoReducer, [], init);

    console.log(todos);

    // Crear la función para manejar el agregado de tareas
    // Mover a TodoAdd y quitar addAction y dispatch
    // const handleSubmit = (e) => {
    //     e.preventDefault();

    //     // Agregar validación para campo vacío luego de hacer funcional el agregado. Le ponemos un mínimo de 3 porque es dificil que hayan tareas de 2 letras. Luego agregamos la validación respectiva en el campo para que notifique al usuario porque no se está agregando
    //     if (description.trim().length < 3) {
    //         return;
    //     }

    //     console.log('Nueva tarea');

    //     // Crear el objeto de la nueva tarea
    //     const newTodo = {
    //         id: new Date().getTime(),
    //         // desc: 'Nueva tarea',

    //         // Asignamos el valor que debe agregar a partir del "bindeo" con el input
    //         // Con esta asignación ya funciona el agregar tarea pero el input no se limpia. Para corregir esto hay que agregar la funcionalidad en el custom hook useForm y así tener centralizadas las funcionalidades
    //         desc: description,
    //         done: false            
    //     };

    //     // Crear la acción de agregar y el valor (payload)
    //     const addAction = {
    //         type: 'add',
    //         payload: newTodo
    //     };

    //     // dispatch realiza la acción del reducer que establecemos en el addAction con el valor del payload
    //     dispatch(addAction);

    //     // Llamamos al reset del custom hook useForm para limpiar el campo una vez hecho el dispatch
    //     // Luego hay que hacer la validación de campo vacío en esta función
    //     handleReset();
    // };

    // const [formValues, handleInputChange] = useForm({
    //     // Recordar que se debe utilizar el nombre (name) del elemento
    //     description: ''
    // });

    // console.log(formValues);

    // const [{ description }, handleInputChange] = useForm({

    // Agregamos el handle para el reset del campo
    // Movemos al componente TodoAdd
    // const [{ description }, handleInputChange, handleReset] = useForm({
    //     // Recordar que se debe utilizar el nombre (name) del elemento
    //     description: ''
    // });

    // console.log(description);

    // Para persistir la información (guardar en el localStorage) podemos observar que el objeto todos (lista de tareas) tenga cambios y para eso podemos utilizar el useEffect
    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos]);

    const handleDelete = (todoId) => {
        console.log(todoId);

        // Crear la acción de eliminar y asignar el payload (valor9), en este caso el id directamente
        const deleteAction = {
            type: 'delete',
            payload: todoId
        };

        // dispatch realiza la acción del reducer que establecemos en el deleteAction con el valor del payload
        // Al refrescar el navegador o revisar el localStorage se puede ver que también se ha eliminado dado que el useEffect esta observando si cambia "todos" y automáticamente actualiza el localStorage
        dispatch(deleteAction);        
    }

    // Para manejar el estado de la tarea.
    // Crear acción en el hook
    const handleToggle = (todoId) => {

        // Forma resumida de llamar al dispatch
        dispatch({
            type: 'toggle',
            payload: todoId
        });        
    };

    // Agregar manejador de agregar todo
    const handleAddTodo = (newTodo) => {
        dispatch({
            type: 'add',
            payload: newTodo
        });        
    };

    return (
        <>
            {/* <h1>To-Do App</h1>   */}

            {/* Agregar la cantidad de tareas */}
            <h1>To-Do App (<small>{todos.length})</small></h1>  
            <hr />

                {/* <li>Comprar pan</li>
                <li>Comprar queso</li>
                <li>Comprar leche</li>
            <li>Comprar jugo</li> */}

                {/* Utilizar bootstrap para dividir en dos columnas la lista y el formulario de agregar */}
                <div className="row">
                    <div className="col-7">

                        {/* Solución Tarea 3, quitar todo el ul y reemplazar por el TodoList */}
                        {/* <ul> */}
                            {
                                // Solución Tarea 1: Agregar condición ternaria para mostrar un mensaje de no hay datos
                                // todos.length > 0 ?
                                //     todos.map((todo, idx) => 
                                        // <li
                                        //     key={todo.id}
                                        //     className="list-group-item"
                                        // >
                                        //     {/* Después de probar el toggle a través de la consola agregar la condición para manejar la clase */}
                                        //     {/* <p  className={todo.done ? 'complete' : ''} onClick={() => handleToggle(todo.id)}>{idx + 1}. { todo.desc }</p> */}

                                        //     {/* Forma corta de manejar la clase */}
                                        //     <p  className={`${todo.done && 'complete'}`} onClick={() => handleToggle(todo.id)}>{idx + 1}. { todo.desc }</p>
                                        //     {/* <button className="btn btn-primary">Borrar</button> */}

                                        //     {/* Agregar el evento onClick para poder eliminar */}
                                        //     <button className="btn btn-danger" onClick={ () => handleDelete(todo.id)}>Borrar</button>
                                        // </li>

                                        // <TodoListItem todo={todo} idx={idx} handleToggle={handleToggle} handleDelete={handleDelete}/>
                                    // )
                                // :
                                //     <h3>No hay tareas</h3>
                            }
                            {/* </ul>  */}
                            
                            <TodoList todos={todos} handleToggle={handleToggle} handleDelete={handleDelete} />
                    </div>

                    {/* <div className="col-5">
                        <h4>Agregar tareas</h4>
                        <hr /> */}

                        {/* Solución Tarea 3, quitar todo el form y reemplazar por el TodoAdd */}
                        {/* <form className="form-group" onSubmit={ handleSubmit }>
                            <input 
                                type="text" 
                                name="description" 
                                placeholder="Ingrese la tarea" 
                                autoComplete="off" 
                                className="form-control" 

                                // Agregamos el manejador del elemento a través del custom hook useForm
                                onChange={ handleInputChange } 

                                // Agregarmos el valor que le corresponde (binding) al elemento. Podría haber sido formValues.description pero por la destructuración queda más sencillo de entender
                                value={ description } 

                                // Agregar para que el usuario sepa que deben ser por lo menos X caracteres
                                minLength="3"
                            />
                            <button type="submit" className="btn btn-outline-primary mt-2 w-100">Agregar</button>
                        </form> */}

                    {/* </div> */}
                    
                    <TodoAdd handleAddTodo={handleAddTodo} />
    
                </div>


        </>
    )
}

// Tarea 1
// Luego de completar el localStorage, agregar un mensaje cuando no haya nada

// Tarea 2
// Implementar el borrado en base a la lógica que se acaba de agregar al reducer para eliminar y este código en el componente:
// const handleDelete = (todoId) => {
//     console.log(todoId);
// }
// Crear la acción
// Crear el dispatch
// Utilizar onClick

// Tarea 3
// Crear un componente TodoList para manejar la lista de tareas (el <ul>). Va a necesitar: todos, handleToggle, handleDelete.
// Crear un componente TodoListItem para manejar cada elemento de la lista de tareas (el <li>). Va a necesitar: todo, idx, handleToggle, handleDelete.

