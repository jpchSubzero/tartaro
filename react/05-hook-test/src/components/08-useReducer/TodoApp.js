import React, { useEffect, useReducer } from 'react';
import {todoReducer} from './todoReducer';
import './styles.css';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';

export const TodoApp = () => {
    const init = () => {
        return JSON.parse(localStorage.getItem('todos')) || [];
    };

    const [todos, dispatch] = useReducer(todoReducer, [], init);

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos]);

    const handleDelete = (todoId) => {
        const deleteAction = {
            type: 'delete',
            payload: todoId
        };

        dispatch(deleteAction);        
    }

    const handleToggle = (todoId) => {
        dispatch({
            type: 'toggle',
            payload: todoId
        });        
    };

    const handleAddTodo = (newTodo) => {
        dispatch({
            type: 'add',
            payload: newTodo
        });        
    };

    return (
        <>
            <h1>To-Do App (<small>{todos.length})</small></h1>  
            <hr />
                <div className="row">
                    <div className="col-7">
                        <TodoList todos={todos} handleToggle={handleToggle} handleDelete={handleDelete} />
                    </div>
                    <TodoAdd handleAddTodo={handleAddTodo} />
                </div>


        </>
    )
}

