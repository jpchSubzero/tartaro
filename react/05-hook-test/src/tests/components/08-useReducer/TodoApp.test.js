import React from 'react';
import { shallow, mount } from 'enzyme';
import { TodoApp } from '../../../components/08-useReducer/TodoApp';
import { mockTodos } from '../../fixtures/mockTodos';
import { act } from "@testing-library/react";

describe('Pruebas en TodoApp', () => {
    const wrapper = shallow(<TodoApp />);

    // Para poder evaluar si se llamó al local storage definimos una función jest para utilizar setItem
    Storage.prototype.setItem = jest.fn(() => {});

    test('Debe mostrarse correctamente', () => {
       expect(wrapper).toMatchSnapshot() ;
    });

    test('Debe agregar un todo', () => {
        // shallow() permite testear un componente sin siquiera pasar por el DOM rendering que hace React. La ventaja de este método es que, al no requerir un entorno de navegador para funcionar, es muy rápido.
        // mount() sí realiza un render en el DOM del componente, lo cual nos permite testear interacciones del usuario (eventos).
        // render() lo que hace es devolver la estructura HTML estática resultante del render del componente, y nos permite navegarla a través de un wrapper de la popular librería cheerio.
        // https://medium.com/@p4bloch/testeando-componentes-de-react-con-enzyme-5aa0468a0cab
        const wrapper = mount(<TodoApp />);

        // Esto nos va a presentar el error: Warning: An update to TodoApp inside a test was not wrapped in act(...). Para solventarlo tenemos que importar act pero desde @testing-library/react ya que estamos probando un componente y no un hook
        // No olvidar que buscamos el componente que utilice la función, en este caso TodoAdd
        // wrapper.find('TodoAdd').prop('handleAddTodo')(mockTodos[0]);
        
        // Una vez agregado act podemos llamar directamente a la función
        act(() => {
            wrapper.find('TodoAdd').prop('handleAddTodo')(mockTodos[0]);
            wrapper.find('TodoAdd').prop('handleAddTodo')(mockTodos[1]);
        });

        // Validar que en el título del componente se vea el mismo número de elementos agregados
        expect(wrapper.find('h1').text().trim()).toBe('To-Do App (2)');

        // Evaluamos si setItem de local storage fue llamada el mismo número de veces que todos agregados
        expect(localStorage.setItem).toHaveBeenCalledTimes(2);
    });

    test('Debe borrarse un todo', () => {
        // Agregar un todo para luego eliminarlo. No olvidar que buscamos el componente que utilice la función, en este caso TodoAdd
        wrapper.find('TodoAdd').prop('handleAddTodo')(mockTodos[0]);
        expect(wrapper.find('h1').text().trim()).toBe('To-Do App (1)');

        // No olvidar que buscamos el componente que utilice la función, en este caso TodoList
        wrapper.find('TodoList').prop('handleDelete')(mockTodos[0].id);
        expect(wrapper.find('h1').text().trim()).toBe('To-Do App (0)');
    });
});
