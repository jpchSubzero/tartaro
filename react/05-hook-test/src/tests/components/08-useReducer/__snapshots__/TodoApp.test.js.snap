// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en TodoApp Debe mostrarse correctamente 1`] = `
<Fragment>
  <h1>
    To-Do App (
    <small>
      0
      )
    </small>
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <TodoAdd
      handleAddTodo={[Function]}
    />
  </div>
</Fragment>
`;
